### PR TITLE
InventoryController

### PR DESCRIPTION
**Pull request by bwu113**
modified all inventory related lines to use inventorycontroller methods.

### CODE REVIEW

File: MapleServer2/Data/Static/ItemMetadataStorage.cs [hunk 1]
```diff
@@ -29,7 +29,7 @@ public static class ItemMetadataStorage {
             return map.GetValueOrDefault(itemId).Slot;
         }
 
-        public static InventoryTab GetTab(int itemId) {
+        public static InventoryType GetTab(int itemId) {
```
Comment (AlanMorel) [inline MapleServer2/Data/Static/ItemMetadataStorage.cs:32]: Is this tab or type? Should this be renamed to `GetType` then?

File: MapleServer2/PacketHandlers/Game/RequestItemInventoryHandler.cs [hunk 2]
```diff
@@ -48,66 +49,31 @@ public override void Handle(GameSession session, PacketReader packet)
...
         }
 
         private void HandleSort(GameSession session, PacketReader packet, Inventory inventory)
         {
-            var tab = (InventoryTab)packet.ReadShort();
-            inventory.Sort(tab);
-
-            session.Send(ItemInventoryPacket.ResetTab(tab));
-            session.Send(ItemInventoryPacket.LoadItemsToTab(tab, inventory.GetItems(tab)));
+            var tab = (InventoryType)packet.ReadShort();
```
Comment (AlanMorel) [inline MapleServer2/PacketHandlers/Game/RequestItemInventoryHandler.cs:75]: Replace `var` with the actual type

File: MapleServer2/Packets/Helpers/ItemPacketHelper.cs [hunk 3]
```diff
@@ -36,7 +36,7 @@ public static class ItemPacketHelper {
                 pWriter.WriteTemplate();
             }
 
-            if (item.InventoryType == InventoryTab.Pets) {
+            if (item.InventoryTab == InventoryType.Pets) {
```
Comment (AlanMorel) [inline MapleServer2/Packets/Helpers/ItemPacketHelper.cs:39]: This whole `Tab` vs `Type` is kind of confusing, maybe we should just pick one and stick with it throughout?

File: MapleServer2/Packets/ItemInventoryPacket.cs [hunk 4]
```diff
@@ -7,13 +7,13 @@
 
 namespace MapleServer2.Packets {
     public static class ItemInventoryPacket {
-        public static Packet ResetTab(InventoryTab tab) {
+        public static Packet ResetTab(InventoryType tab) {
```
Comment (AlanMorel) [inline MapleServer2/Packets/ItemInventoryPacket.cs:10]: Same here, is it type or tab?

File: MapleServer2/Tools/InventoryController.cs [hunk 5]
```diff
@@ -1,46 +1,153 @@
...
-        session.Player.Inventory.Add(item); 
-        session.Send(ItemInventoryPacket.Add(item));
-        session.Send(ItemInventoryPacket.MarkItemNew(item, item.Amount));
+        session.Player.Inventory.Add(item); // adds item into internal database
+        session.Send(ItemInventoryPacket.Add(item)); // sends packet to add item clientside.
+
+        if (isNew)
+        {
+            session.Send(ItemInventoryPacket.MarkItemNew(item, item.Amount)); // Marks Item as New
+        }
```
Comment (rhydev) [inline MapleServer2/Tools/InventoryController.cs:45]: This change to how MarkItemNew is handled is wrong and also adds a needless parameter to the function, please revert. With this implementation, stacks won't be marked as new.

File: MapleServer2/Tools/InventoryController.cs [hunk 6]
```diff
@@ -1,46 +1,153 @@
...
                 {
                     continue;
                 }
-                // Updates inventory for item amount overflow
+                // Updates inventory for item amount overflow.
                 if ((i.Amount + item.Amount) > i.SlotMax)
                 {
-                    int added = i.SlotMax - i.Amount; // For marking item new with correct added amount
-
                     item.Amount = item.Amount - (i.SlotMax - i.Amount);
```
Comment (AlanMorel) [inline MapleServer2/Tools/InventoryController.cs:26]: Could simplify this to using the `-=` operator

File: MapleServer2/Tools/InventoryController.cs [hunk 7]
```diff
@@ -1,46 +1,153 @@
...
                     i.Amount = i.SlotMax;
                     session.Send(ItemInventoryPacket.Update(i.Uid, i.Amount));
-                    session.Send(ItemInventoryPacket.MarkItemNew(i, added));
                 }
                 // Updates item amount
                 else
                 {
                     i.Amount = i.Amount + item.Amount;
```
Comment (AlanMorel) [inline MapleServer2/Tools/InventoryController.cs:33]: Could simplify this to using the `+=` operator

File: MapleServer2/Tools/InventoryController.cs [hunk 8]
```diff
@@ -1,46 +1,153 @@
...
             foreach (Item i in session.Player.Inventory.Items.Values)
             {
-                // Continue if inventory item id doesn't match or it is at max amount
+                // Checks to see if item exists in database (dictionary)
                 if (i.Id != item.Id || i.Amount >= i.SlotMax)
                 {
                     continue;
                 }
-                // Updates inventory for item amount overflow
+                // Updates inventory for item amount overflow.
```
Comment (rhydev) [inline MapleServer2/Tools/InventoryController.cs:23]: Please follow proper comment formatting: space after `//` no period, start with a capital letter, this is a problem throughout the PR

File: MapleServer2/Tools/InventoryController.cs [hunk 9]
```diff
@@ -1,46 +1,153 @@
...
+
+    // Removes Item from inventory by reference
+    public static void Remove(GameSession session, out Item item, long uid)
+    {
+        session.Player.Inventory.Remove(uid, out item);
+        session.Send(ItemInventoryPacket.Remove(uid));
+    }
+
+    // Removes Item from inventory
+    public static void Remove(GameSession session, Item item, long uid)
```
Comment (AlanMorel) [inline MapleServer2/Tools/InventoryController.cs:56]: This seems identical to the method above

File: Maple2Storage/Types/InventoryType.cs [hunk 10]
```diff
@@ -1,5 +1,5 @@
 ﻿namespace Maple2Storage.Types {
-    public enum InventoryTab : byte {
+    public enum InventoryType : byte {
```
Comment (rhydev) [inline Maple2Storage/Types/InventoryType.cs:2]: Why was this changed?

File: MapleServer2/Tools/InventoryController.cs [hunk 11]
```diff
@@ -1,46 +1,153 @@
...
+        if (!isbound) // Drops Item.
+        {
+            int remaining = session.Player.Inventory.Remove(uid, out Item droppedItem, amount); // returns remaining amount of item.
+            if (remaining < 0)
+            {
+                return; // Removal failed
+            }
+
+            // Updates item amount and removes item.
+            session.Send(remaining > 0
```
Comment (AlanMorel) [inline MapleServer2/Tools/InventoryController.cs:81]: can this be broken out into an if/else block?

File: MapleServer2/PacketHandlers/Game/RequestItemUseHandler.cs [hunk 12]
```diff
@@ -28,19 +28,18 @@ public override void Handle(GameSession session, PacketReader packet)
             // Remove box if amount is 1 or less
             if (box.Amount <= 1)
             {
-                session.Player.Inventory.Remove(boxUid, out Item removed);
-                session.Send(ItemInventoryPacket.Remove(boxUid));
+                InventoryController.Remove(session, out Item removed, boxUid);
```
Comment (rhydev) [inline MapleServer2/PacketHandlers/Game/RequestItemUseHandler.cs:31]: `out` should be last parameter

**Review by AlanMorel — CHANGES_REQUESTED** [review]

File: MapleServer2/Tools/InventoryController.cs [hunk 13]
```diff
@@ -1,46 +1,153 @@
...
+    // Sorts inventory items
+    public static void SortInventory(GameSession session, Inventory inventory, InventoryType tab)
+    {
+        inventory.Sort(tab);
+        session.Send(ItemInventoryPacket.ResetTab(tab));
+        session.Send(ItemInventoryPacket.LoadItemsToTab(tab, inventory.GetItems(tab)));
+    }
+
+    // Loads all inventory tabs
+    public static void LoadInventoryTabs(GameSession session, InventoryType tab)
```
Comment (rhydev) [inline MapleServer2/Tools/InventoryController.cs:108]: This seems to load a single tab but it's made to seem like it loads all tabs?

File: MapleServer2/Tools/InventoryController.cs [hunk 14]
```diff
@@ -1,46 +1,153 @@
...
+    public static void PickUp(GameSession session, Item item)
+    {
+        session.Player.Inventory.Add(item); // adds item into internal database
+        session.Send(ItemInventoryPacket.Add(item)); // sends packet to add item clientside.
+    }
+
+    // Drops item with option to drop bound items.
+    public static void DropItem(GameSession session, long uid, int amount, Boolean isbound)
+    {
+        if (!isbound) // Drops Item.
```
Comment (rhydev) [inline MapleServer2/Tools/InventoryController.cs:72]: Should this really be a parameter?

File: MapleServer2/Tools/InventoryController.cs (hunk 11 above)
Comment (rhydev) [inline MapleServer2/Tools/InventoryController.cs:81]: Ternary operator shouldn't be used here since it's sending entirely different packets

**Review by rhydev — CHANGES_REQUESTED** [review]

File: Maple2Storage/Types/InventoryType.cs (hunk 10 above)
Comment (newgithubaccountlol) [reply Maple2Storage/Types/InventoryType.cs:2]: I think we should just keep the old one. It's pretty self explanatory what it is and it's already used in many files. If this needs to be changed, I'd say we should make another PR that only has the refactoring changes.

File: MapleServer2/Enums/ItemRarity.cs [hunk 15]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace MapleServer2.Enums
+{
+    public enum RarityType
```
Comment (newgithubaccountlol) [inline MapleServer2/Enums/ItemRarity.cs:7]: is this used anywhere? or is it just placeholder? If this is used, are these the same numbers nexon uses?

File: MapleServer2/PacketHandlers/Game/FieldEnterHandler.cs [hunk 16]
```diff
@@ -57,9 +57,9 @@ public override void Handle(GameSession session, PacketReader packet)
             {
                 Amount = 10000
             };
-            InventoryController.Add(session, item);
-            InventoryController.Add(session, item2);
-            InventoryController.Add(session, item3);
+            InventoryController.Add(session, item, true);
```
Comment (newgithubaccountlol) [inline MapleServer2/PacketHandlers/Game/FieldEnterHandler.cs:60]: Should these be false? We don't need new notifs for this every time we enter.

File: MapleServer2/Tools/InventoryController.cs [hunk 17]
```diff
@@ -1,44 +1,151 @@
...
                     continue;
                 }
-                // Updates inventory for item amount overflow
+                // Updates inventory for item amount overflow.
                 if ((i.Amount + item.Amount) > i.SlotMax)
                 {
-                    int added = i.SlotMax - i.Amount; // For marking item new with correct added amount
-
-                    item.Amount -= added;
+                    item.Amount -= (i.SlotMax - i.Amount);
```
Comment (newgithubaccountlol) [inline MapleServer2/Tools/InventoryController.cs:24]: [nit] parens not needed and don't help readability

**Review by newgithubaccountlol — COMMENTED** [review]
Looks pretty good to me. I agree with the other comments so make sure to address them

File: MapleServer2/Enums/ItemRarity.cs (hunk 15 above)
Comment (bwu113) [reply MapleServer2/Enums/ItemRarity.cs:7]: These aren't being used atm. But will be later on. and yes these are the same numbers nexon uses.

File: MapleServer2/PacketHandlers/Game/RequestItemInventoryHandler.cs [hunk 18]
```diff
@@ -31,8 +31,9 @@ public override void Handle(GameSession session, PacketReader packet)
             switch (mode)
             {
                 case RequestItemInventoryMode.Move:
-                    HandleMove(session, packet, inventory);
+                    HandleMove(session, packet);
                     break;
+
```
Comment (AlanMorel) [inline MapleServer2/PacketHandlers/Game/RequestItemInventoryHandler.cs:36]: This blank line isn't needed

File: MapleServer2/Tools/InventoryController.cs [hunk 19]
```diff
@@ -36,9 +38,110 @@ public static void Add(GameSession session, Item item)
...
+        if (!isbound) // Drops Item
+        {
+            int remaining = session.Player.Inventory.Remove(uid, out Item droppedItem, amount); // Returns remaining amount of item
+            if (remaining < 0)
+            {
+                return; // Removal failed
+            }
+
+            // Updates item amount and removes item
+            session.Send(remaining > 0
```
Comment (AlanMorel) [inline MapleServer2/Tools/InventoryController.cs:75]: Can this be turned into an if/else?

**Review by AlanMorel — CHANGES_REQUESTED** [review]
